### PR TITLE
[Feature][Executor] Support line timeout in exec_line to ensure run_info is also persisted in batch run.

### PR DESCRIPTION
**Pull request by thy09**
# Description

This pull request introduces several changes to improve the handling of timeouts and execution duration in the codebase. The most important changes include adding a new parameter to specify a timeout duration for each line of execution in the `FlowNodesScheduler` class, adding a method to handle timeouts and skip further processing for canceled runs in the `RunTracker` class, and making the line timeout value optional and retrievable from an environment variable in the `FlowExecutor` and `_line_execution_process_pool` classes. This PR also adjusts the logic in process pool to collect such canceled node runs so the node statuses could be correct.

Main changes:
* [`src/promptflow/promptflow/executor/_flow_nodes_scheduler.py`](diffhunk://#diff-33de607e09569e7e33422ca25579a82d0418874a4e917cd711b48dc8c6dd585cR8-R19): Added the `Optional` type to the `typing` import and added a `wait_within_timeout` method to handle line execution timeouts. Modified the `execute` method to accept an optional `line_timeout_sec` parameter and use another thread to monitor the timeout. [[1]](diffhunk://#diff-33de607e09569e7e33422ca25579a82d0418874a4e917cd711b48dc8c6dd585cR8-R19) [[2]](diffhunk://#diff-33de607e09569e7e33422ca25579a82d0418874a4e917cd711b48dc8c6dd585cR42-R94)
* `src/promptflow/promptflow/executor/_line_execution_process_pool.py`: Refactored code to avoid duplicating keyword arguments, added logging statements and error handling for line execution timeouts and crashes. (F797de53)
* [`src/promptflow/promptflow/_core/flow_execution_context.py`](diffhunk://#diff-8a45b6238b72974b62aa211aec63ef4cbeadfa8277f84525442c245a16ee4461R70-R72): Added a new method `cancel_node_runs` to the `FlowExecutionContext` class.
* [`src/promptflow/promptflow/executor/_process_manager.py`](diffhunk://#diff-fb203fa9d75078068758d8934fd7711b63934b15faa6cee7c3c4d16ab3347258L384-R391): Removed unnecessary arguments from various methods and constructors, and instead passed a `flow_create_kwargs` dictionary. [[1]](diffhunk://#diff-fb203fa9d75078068758d8934fd7711b63934b15faa6cee7c3c4d16ab3347258L384-R391) [[2]](diffhunk://#diff-fb203fa9d75078068758d8934fd7711b63934b15faa6cee7c3c4d16ab3347258L55-L62) [[3]](diffhunk://#diff-fb203fa9d75078068758d8934fd7711b63934b15faa6cee7c3c4d16ab3347258L226-R222) [[4]](diffhunk://#diff-fb203fa9d75078068758d8934fd7711b63934b15faa6cee7c3c4d16ab3347258L420) [[5]](diffhunk://#diff-fb203fa9d75078068758d8934fd7711b63934b15faa6cee7c3c4d16ab3347258L206-R207)
* [`src/promptflow/promptflow/_core/run_tracker.py`](diffhunk://#diff-3ea41c85102f8318e1dc8568de0fb0de76ce92172e506df95e6f419808eabaf8R231-R245): Added a new method to cancel node runs, handled canceled runs in the `end_run` method, and imported the `ToolCancelledError` class. [[1]](diffhunk://#diff-3ea41c85102f8318e1dc8568de0fb0de76ce92172e506df95e6f419808eabaf8R231-R245) [[2]](diffhunk://#diff-3ea41c85102f8318e1dc8568de0fb0de76ce92172e506df95e6f419808eabaf8R264-R266) [[3]](diffhunk://#diff-3ea41c85102f8318e1dc8568de0fb0de76ce92172e506df95e6f419808eabaf8L12-R12)
* [`src/promptflow/promptflow/executor/flow_executor.py`](diffhunk://#diff-faa6c81d614b7e41b18a42a93139d961d92afa9aa9dd0b72cb6b7176d7541e69L114-R114): Updated the type annotation for the `line_timeout_sec` parameter, removed the import of the `LINE_TIMEOUT_SEC` constant, and added the `get_int_env_var` function to the import statement. [[1]](diffhunk://#diff-faa6c81d614b7e41b18a42a93139d961d92afa9aa9dd0b72cb6b7176d7541e69L114-R114) [[2]](diffhunk://#diff-faa6c81d614b7e41b18a42a93139d961d92afa9aa9dd0b72cb6b7176d7541e69L16-R16) [[3]](diffhunk://#diff-faa6c81d614b7e41b18a42a93139d961d92afa9aa9dd0b72cb6b7176d7541e69L38-R38)
* [`src/promptflow/tests/sdk_cli_test/conftest.py`](diffhunk://#diff-9dced3762e0027d9b6b131576702c5290a82d72c20b2a828c67aa682d31d8e40L5): Removed the import statement for the `Queue` class and modified methods to accept variable arguments and keyword arguments. [[1]](diffhunk://#diff-9dced3762e0027d9b6b131576702c5290a82d72c20b2a828c67aa682d31d8e40L5) [[2]](diffhunk://#diff-9dced3762e0027d9b6b131576702c5290a82d72c20b2a828c67aa682d31d8e40L258-R264)

Other changes:

* [`src/promptflow/tests/executor/e2etests/test_batch_timeout.py`](diffhunk://#diff-f3f345ba5efdfd8e3d9ad093ae9c2a0639925c41f9b90d5376db7dadf1943291L32-R68): Replaced a parameter, added assertions, updated error messages, and changed timeout values. [[1]](diffhunk://#diff-f3f345ba5efdfd8e3d9ad093ae9c2a0639925c41f9b90d5376db7dadf1943291L32-R68) [[2]](diffhunk://#diff-f3f345ba5efdfd8e3d9ad093ae9c2a0639925c41f9b90d5376db7dadf1943291R108-R112) [[3]](diffhunk://#diff-f3f345ba5efdfd8e3d9ad093ae9c2a0639925c41f9b90d5376db7dadf1943291L58-R77) [[4]](diffhunk://#diff-f3f345ba5efdfd8e3d9ad093ae9c2a0639925c41f9b90d5376db7dadf1943291L74-R95) [[5]](diffhunk://#diff-f3f345ba5efdfd8e3d9ad093ae9c2a0639925c41f9b90d5376db7dadf1943291L98-R127)
* [`src/promptflow/tests/executor/utils.py`](diffhunk://#diff-26387b75113dc30aff56f241631293fb04db301f61c8d4a957c4b80bd9e17cfdL104-R105): Updated type annotations in the `MemoryRunStorage` class and the `is_image_file` function. [[1]](diffhunk://#diff-26387b75113dc30aff56f241631293fb04db301f61c8d4a957c4b80bd9e17cfdL104-R105) [[2]](diffhunk://#diff-26387b75113dc30aff56f241631293fb04db301f61c8d4a957c4b80bd9e17cfdL3-R3)
* [`src/promptflow/tests/test_configs/flows/one_line_of_bulktest_timeout/samples_all_timeout.json`](diffhunk://#diff-ce4fc7e790810972768df270ea6406c357b720bacd205d4156868a244af34ca9R1): Added a new line to the `samples_all_timeout.json` file.
* [`src/promptflow/promptflow/executor/_errors.py`](diffhunk://#diff-1f02407850413baccbb33e0078d9772e15ae1c46c020dc7457529c6f7a054315R173): Added a new attribute `timeout_sec` to the `LineExecutionTimeoutError` class.
* [`src/promptflow/promptflow/_core/_errors.py`](diffhunk://#diff-d7c75a55b795a9cdff3801c0911a2cbd102961d8c1f7031aaf36169e1701f9c9R40-R44): Added a new exception class `ToolCancelledError`.

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/_core/_errors.py [hunk 1]
```diff
@@ -37,6 +37,11 @@ class InputTypeMismatch(ValidationException):
     pass
 
 
+class ToolCancelledError(UserErrorException):
```
Comment (zzn2) [inline src/promptflow/promptflow/_core/_errors.py:40]: `Cancelled` or `Canceled`?

**Review by zzn2 — APPROVED** [review]

File: src/promptflow/promptflow/executor/_line_execution_process_pool.py [hunk 2]
```diff
@@ -292,12 +284,26 @@ def _monitor_workers_and_process_tasks_in_thread(
                 log_process_status(process_name, process_id, line_number, is_completed=True)
             # Handle line execution is not completed.
             else:
+                ex = None
                 # Handle process crashed.
                 if crashed:
+                    bulk_logger.warning(f"Process crashed while executing line {line_number}.")
+                    ex = ProcessCrashError(line_number)
                     self.handle_process_crashed(line_number, inputs, run_id, start_time, result_list)
```
Comment (Hhhilulu) [inline src/promptflow/promptflow/executor/_line_execution_process_pool.py:292]: Need to be deleted?

File: src/promptflow/promptflow/_core/_errors.py (hunk 1 above)
Comment (thy09) [reply src/promptflow/promptflow/_core/_errors.py]: Updated to canceled

File: src/promptflow/promptflow/executor/_line_execution_process_pool.py (hunk 2 above)
Comment (thy09) [reply src/promptflow/promptflow/executor/_line_execution_process_pool.py]: Good catch, removed.

Comment (thy09) [reply src/promptflow/promptflow/executor/_line_execution_process_pool.py]: Do we have a test to cover this?

**Review by crazygao — APPROVED** [review]

**Review by zzn2 — APPROVED** [review]

**Pull request merged**